test/VaultInventoryReporter.ts

```diff
@@ -0,0 +1,1063 @@
+                    .withArgs(vault.address, user.address, hashItem(items[1]))
+                    .to.emit(reporter, "Add")
+                    .withArgs(vault.address, user.address, hashItem(items[2]))
+                    .to.emit(reporter, "Add")
+                    .withArgs(vault.address, user.address, hashItem(items[3]));
+            });
+
+            it("should allow an approved address to add items to inventory", async () => {
+                const { reporter, user, other, vault } = ctx;
+
+                await reporter.connect(user).setApproval(vault.address, other.address);
+
+                const items = defaultItems;
+
+                await expect(
+                    reporter.connect(other).add(vault.address, items)
+                ).to.emit(reporter, "Add")
+                    .withArgs(vault.address, other.address, hashItem(items[0]))
+                    .to.emit(reporter, "Add")
+                    .withArgs(vault.address, other.address, hashItem(items[1]))
+                    .to.emit(reporter, "Add")
+                    .withArgs(vault.address, other.address, hashItem(items[2]))
+                    .to.emit(reporter, "Add")
+                    .withArgs(vault.address, other.address, hashItem(items[3]));
+
+                expect((await reporter.keys(vault.address)).length).to.eq(4);
+            });
+
+            it("should not allow an address to add items to inventory with an invalid permit signature", async () => {
+                const { reporter, user, other, vault } = ctx;
+
+                const deadline = maxDeadline;
+
+                const permitData: InventoryPermitData = {
+                    owner: user.address,
+                    target: other.address,
+                    vault: vault.address,
+                    nonce: 1000, // use invalid nonce
+                    deadline
+                };
+
+                const sig = await createInventoryPermitSignature(
+                    reporter.address,
+                    NAME,
+                    permitData,
+                    user
+                );
+
+                const items = defaultItems;
+
+                await expect(
+                    reporter.connect(other).addWithPermit(vault.address, items, deadline, sig.v, sig.r, sig.s)
+                ).to.be.revertedWith("VIR_InvalidPermitSignature");
+            });
+
+            it("should not allow an address to add items to inventory with an expired permit signature", async () => {
+                const { reporter, user, other, vault } = ctx;
+
+                // Make deadline in the past
+                const lastBlock = await hre.ethers.provider.getBlock("latest");
+                const deadline = lastBlock.timestamp - 1000;
+
+                const permitData: InventoryPermitData = {
+                    owner: user.address,
+                    target: other.address,
+                    vault: vault.address,
+                    nonce: 0,
```

Comment:
Obviously the test is passing, but shouldn't a nonce with value zero revert?
`nonce: 0` is also on lines 431, 592, 619, 744, 769.

Should a test be added for potential signature nonce re-use?

Reply:
A nonce with value 0 is allowed here, remember this is not our _own_ signature protocol, but the standard `Permit` protocol. For `ERC721Permit`, the nonces used must be serial and start with 0 (see lines 554-558).

Additional test sounds like a great idea, I'll add.